urldeb33fd2fa_friedemann_distripedia_tgz-pJ8-org_json_TestJ8/src/grinder/client/Client.java

```diff
@@ -0,0 +1,512 @@
+package grinder.client;
+
+import grinder.client.user.BrowserType;
+import grinder.client.user.User;
+import grinder.util.GrinderSettings;
+import grinder.util.Randomizer;
+import grinder.util.TestTicket;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.Socket;
+import java.net.UnknownHostException;
+import java.util.Enumeration;
+import java.util.Hashtable;
+import java.util.Properties;
+import java.util.UUID;
+
+/**
+ * The Client is first controlled by the client gui until a server connection is
+ * set up. From then on the server instance controls the client, and the client
+ * only reports back to the gui. Of course the gui can always shut down its
+ * client, but that could result in a complete loss of data (!).
+ *
+ * @author Friedemann
+ */
+public class Client {
+
+	/** The client socket. */
+	private Socket clientSocket;
+
+	/** The unique client id. */
+	private final UUID clientID;
+
+	/** The controller/server ip. */
+	private InetAddress controllerIP;
+
+	/** The current client status . */
+	private ClientState state;
+
+	/** True as long as the client is listening for incoming commands */
+	private boolean listening = true;
+
+	/** The object outsourcing the client/controller connection. */
+	private ClientConnectionThread cct;
+
+	/** The thread controlling the actual grinding. */
+	private GrinderThread gt;
+
+	/** true if a server connection is established. */
+	public boolean connectionEstablished = false;
+
+	/** The Client GUI the Client itself can talk to. */
+	public ClientGUI cgui;
+
+	/** The Testcase data received from the server. */
+	public TestTicket testTicket;
+
+	/** Storage for the active users, http type */
+	public Hashtable<UUID, User> userHTTP;
+
+	/** Storage for the active users, websocket type */
+	public Hashtable<UUID, User> userWS;
+
+	/** Storage for the active users, stratus type */
+	public Hashtable<UUID, User> userSTRATUS;
+
+	/** The client's logfiler */
+	public LogFiler lf;
+
+	private boolean allUsersInitialized;
+
+	private Properties props;
+
+
+	/**
+	 * Instantiates a new client.
+	 *
+	 * @param clientGUI the client gui
+	 */
+	public Client(final ClientGUI clientGUI) {
+
+		cgui = clientGUI;
+		clientID = UUID.randomUUID();
+		state = ClientState.IDLE;
+		
+		userHTTP = new Hashtable<UUID, User>(50);
+		userWS = new Hashtable<UUID, User>(50);
+		userSTRATUS = new Hashtable<UUID, User>(50);
+
+		loadProperties("client.properties");
+		cgui.logMessage("...initialized. id: " + getClientID());
+	}
+
+	/**
+	 * Load properties.
+	 *
+	 * @param propFileName the prop file name
+	 * @return the properties
+	 */
+	private Properties loadProperties(final String propFileName) {
+
+		final Properties p = new Properties();
+
+		try
+		{
+			final FileInputStream stream = new FileInputStream(propFileName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *stream*. The resource is referenced at line 110. The resource is closed at line 111. There are other execution paths that do not contain closure statements, for example, when Properties.load throws an exception. Ensure *stream* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).